rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -320,111 +454,283 @@ rmw_create_publisher(
+}
+
+rmw_ret_t
+rmw_destroy_publisher(rmw_node_t * node, rmw_publisher_t * publisher)
+{
+  if (!node) {
+    rmw_set_error_string("node handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (node->implementation_identifier != rti_connext_dynamic_identifier) {
+    rmw_set_error_string("node handle is not from this rmw implementation");
+    return RMW_RET_ERROR;
+  }
+  DDSDomainParticipant * participant = (DDSDomainParticipant *)node->data;
+  if (!participant) {
+    rmw_set_error_string("participant handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (!publisher) {
+    rmw_set_error_string("publisher handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (publisher->implementation_identifier != rti_connext_dynamic_identifier) {
+    rmw_set_error_string("publisher handle is not from this rmw implementation");
+    return RMW_RET_ERROR;
+  }
+  CustomPublisherInfo * custom_publisher_info = (CustomPublisherInfo *)publisher->data;
+  if (custom_publisher_info) {
+    DDSDynamicDataTypeSupport * ddts = custom_publisher_info->dynamic_data_type_support_;
+    if (ddts) {
+      std::string type_name = _create_type_name(custom_publisher_info->members_, "msg");
+      if (ddts->unregister_type(participant, type_name.c_str()) != DDS_RETCODE_OK) {
+        rmw_set_error_string(("failed to unregister type: " + type_name).c_str());
+        return RMW_RET_ERROR;
+      }
+      if (custom_publisher_info->dynamic_data) {
+        if (ddts->delete_data(custom_publisher_info->dynamic_data) != DDS_RETCODE_OK) {
+          rmw_set_error_string("failed to delete dynamic data");
+          return RMW_RET_ERROR;
+        }
+      }
+      custom_publisher_info->dynamic_data = nullptr;
+      rmw_free(ddts);
+    }
+    custom_publisher_info->dynamic_data_type_support_ = nullptr;
+    DDSPublisher * dds_publisher = custom_publisher_info->dds_publisher_;
+    if (dds_publisher) {
+      auto data_writer = custom_publisher_info->data_writer_;
+      if (data_writer) {
+        if (dds_publisher->delete_datawriter(data_writer) != DDS_RETCODE_OK) {
+          rmw_set_error_string("failed to delete datawriter");
+          return RMW_RET_ERROR;
+        }
+      }
+      custom_publisher_info->data_writer_ = nullptr;
+      custom_publisher_info->dynamic_writer_ = nullptr;
+      if (participant->delete_publisher(dds_publisher) != DDS_RETCODE_ERROR) {
+        rmw_set_error_string("failed to delete publisher");
+        return RMW_RET_ERROR;
+      }
+    }
+    custom_publisher_info->dds_publisher_ = nullptr;
+    if (custom_publisher_info->type_code_) {
+      if (destroy_type_code(custom_publisher_info->type_code_) != RMW_RET_OK) {
+        // Error string already set.
+        return RMW_RET_ERROR;
+      }
+    }
+    custom_publisher_info->type_code_ = nullptr;
+    rmw_free(custom_publisher_info);
+  }
+  publisher->data = nullptr;
+  rmw_publisher_free(publisher);
+  return RMW_RET_OK;
 }
 
 #define SET_PRIMITIVE_VALUE(TYPE, METHOD_NAME) \
```

Comment:
I prefer macros to be at the top of files, but that's just a personal preference.